cpp/src/arrow/compute/kernels/nth_to_indices.cc

```diff
@@ -0,0 +1,205 @@
+
+    std::iota(indices_begin, indices_end, 0);
+    *offsets = std::make_shared<UInt64Array>(values->length(), indices_buf);
+
+    if (n >= static_cast<uint64_t>(values->length())) {
+      return Status::OK();
+    }
+
+    auto nulls_begin = indices_end;
+    if (values->null_count()) {
+      nulls_begin =
+          std::stable_partition(indices_begin, indices_end,
+                                [&values](uint64_t ind) { return !values->IsNull(ind); });
+    }
+
+    auto nth_begin = indices_begin + n;
+    if (nth_begin < nulls_begin) {
+      std::nth_element(indices_begin, nth_begin, nulls_begin,
+                       [&values, this](uint64_t left, uint64_t right) {
+                         return compare_(*values, left, right);
+                       });
+    }
+    return Status::OK();
+  }
+};
+
+template <typename ArrowType, typename Comparator>
+NthToIndicesKernelImpl<ArrowType, Comparator>* MakeNthToIndicesKernelImpl(
+    Comparator comparator) {
+  return new NthToIndicesKernelImpl<ArrowType, Comparator>(comparator);
+}
+
+Status NthToIndicesKernel::Make(const std::shared_ptr<DataType>& value_type,
+                                std::unique_ptr<NthToIndicesKernel>* out) {
+  NthToIndicesKernel* kernel;
+  switch (value_type->id()) {
+    case Type::UINT8:
+      kernel = MakeNthToIndicesKernelImpl<UInt8Type>(CompareValues<UInt8Array>);
+      break;
+    case Type::INT8:
+      kernel = MakeNthToIndicesKernelImpl<Int8Type>(CompareValues<Int8Array>);
+      break;
+    case Type::UINT16:
+      kernel = MakeNthToIndicesKernelImpl<UInt16Type>(CompareValues<UInt16Array>);
+      break;
+    case Type::INT16:
+      kernel = MakeNthToIndicesKernelImpl<Int16Type>(CompareValues<Int16Array>);
+      break;
+    case Type::UINT32:
+      kernel = MakeNthToIndicesKernelImpl<UInt32Type>(CompareValues<UInt32Array>);
+      break;
+    case Type::INT32:
+      kernel = MakeNthToIndicesKernelImpl<Int32Type>(CompareValues<Int32Array>);
+      break;
+    case Type::UINT64:
+      kernel = MakeNthToIndicesKernelImpl<UInt64Type>(CompareValues<UInt64Array>);
+      break;
+    case Type::INT64:
+      kernel = MakeNthToIndicesKernelImpl<Int64Type>(CompareValues<Int64Array>);
+      break;
+    case Type::FLOAT:
+      kernel = MakeNthToIndicesKernelImpl<FloatType>(CompareValues<FloatArray>);
+      break;
+    case Type::DOUBLE:
+      kernel = MakeNthToIndicesKernelImpl<DoubleType>(CompareValues<DoubleArray>);
+      break;
+    case Type::BINARY:
+      kernel = MakeNthToIndicesKernelImpl<BinaryType>(CompareViews<BinaryArray>);
+      break;
+    case Type::STRING:
+      kernel = MakeNthToIndicesKernelImpl<StringType>(CompareViews<StringArray>);
+      break;
+    default:
+      return Status::NotImplemented("N-th of ", *value_type, " arrays");
+  }
+  out->reset(kernel);
+  return Status::OK();
+}
+
+Status NthToIndices(FunctionContext* ctx, const Datum& values, uint64_t n,
```

Comment:
Do you want to expose this function in the `.h`?

Reply:
no, changed to static.